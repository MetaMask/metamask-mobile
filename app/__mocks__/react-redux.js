export * from 'react-redux-test';
