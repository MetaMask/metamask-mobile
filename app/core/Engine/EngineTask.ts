import Realm from 'realm';

/**
 * Generic schema for persisting any controller state
 * Replaces FileSystemStorage key-value approach with Realm storage
 */
export class PersistedState extends Realm.Object<PersistedState> {
  key!: string;        // "persist:TransactionController", "persist:KeyringController", etc.
  data!: any;        // The actual controller state as JSON-serializable data
  updatedAt!: Date;    // When this controller state was last updated
  version!: number;    // Schema version for future migrations

  static schema: Realm.ObjectSchema = {
    name: 'PersistedState',
    primaryKey: 'key',
    properties: {
      key: 'string',
      data: 'mixed',
      updatedAt: 'date',
      version: { type: 'int', default: 1 },
    },
  };
}

// Keep the old Controller for backward compatibility during transition
export class Controller extends Realm.Object<Controller> {
  key!: string;
  value!: string;
  isActive!: boolean;

  static schema: Realm.ObjectSchema = {
    name: 'Controller',
    primaryKey: 'key',
    properties: {
      key: 'string',
      value: 'string',
      isActive: { type: 'bool', default: true },
    },
  };
}
