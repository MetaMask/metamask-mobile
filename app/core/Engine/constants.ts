/**
 * Messageable modules that are part of the Engine's context, but are not defined with state.
 * TODO: Replace with type guard once consistent inheritance for non-controllers is implemented. See: https://github.com/MetaMask/decisions/pull/41
 */
export const STATELESS_NON_CONTROLLER_NAMES = [
  'AssetsContractController',
  'NftDetectionController',
  'TokenDetectionController',
] as const;

export const BACKGROUND_STATE_CHANGE_EVENT_NAMES = [
  'AccountsController:stateChange',
  'AccountTrackerController:stateChange',
  'AddressBookController:stateChange',
  'ApprovalController:stateChange',
  'CurrencyRateController:stateChange',
  'GasFeeController:stateChange',
  'KeyringController:stateChange',
  'LoggingController:stateChange',
  'NetworkController:stateChange',
  'NftController:stateChange',
  'PermissionController:stateChange',
  'PhishingController:stateChange',
  'PPOMController:stateChange',
  'PreferencesController:stateChange',
  'RemoteFeatureFlagController:stateChange',
  'SelectedNetworkController:stateChange',
  'SignatureController:stateChange',
  'SmartTransactionsController:stateChange',
  'SwapsController:stateChange',
  'TokenBalancesController:stateChange',
  'TokenListController:stateChange',
  'TokenRatesController:stateChange',
  'TokensController:stateChange',
  'TransactionController:stateChange',
  ///: BEGIN:ONLY_INCLUDE_IF(preinstalled-snaps,external-snaps)
  'SnapController:stateChange',
  'SnapsRegistry:stateChange',
  'SubjectMetadataController:stateChange',
  'AuthenticationController:stateChange',
  'UserStorageController:stateChange',
  'NotificationServicesController:stateChange',
  'NotificationServicesPushController:stateChange',
  ///: END:ONLY_INCLUDE_IF
] as const;
