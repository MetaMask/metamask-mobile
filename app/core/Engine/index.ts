export { default } from './Engine';
export * from './types';
