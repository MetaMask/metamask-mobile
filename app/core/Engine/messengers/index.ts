import { getAccountsControllerMessenger } from './accounts-controller-messenger';
import type { ControllerMessengerByControllerName } from '../types';
import { getMultichainNetworkControllerMessenger } from './multichain-network-controller-messenger/multichain-network-controller-messenger';
import { getCurrencyRateControllerMessenger } from './currency-rate-controller-messenger/currency-rate-controller-messenger';
///: BEGIN:ONLY_INCLUDE_IF(preinstalled-snaps,external-snaps)
import { getCronjobControllerMessenger } from './cronjob-controller-messenger/cronjob-controller-messenger';
///: END:ONLY_INCLUDE_IF
///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
import { getMultichainAssetsRatesControllerMessenger } from './multichain-assets-rates-controller-messenger/multichain-assets-rates-controller-messenger';
import { getMultichainAssetsControllerMessenger } from './multichain-assets-controller-messenger/multichain-assets-controller-messenger';
import { getMultichainBalancesControllerMessenger } from './multichain-balances-controller-messenger/multichain-balances-controller-messenger';
///: END:ONLY_INCLUDE_IF

/**
 * The messengers for the controllers that have been.
 */
export const CONTROLLER_MESSENGERS: ControllerMessengerByControllerName = {
  AccountsController: {
    getMessenger: getAccountsControllerMessenger,
  },
  CurrencyRateController: {
    getMessenger: getCurrencyRateControllerMessenger,
  },
  MultichainNetworkController: {
    getMessenger: getMultichainNetworkControllerMessenger,
  },
  ///: BEGIN:ONLY_INCLUDE_IF(preinstalled-snaps,external-snaps)
  CronjobController: {
    getMessenger: getCronjobControllerMessenger,
  },
  ///: END:ONLY_INCLUDE_IF
  ///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
  MultichainAssetsController: {
    getMessenger: getMultichainAssetsControllerMessenger,
  },
  MultichainAssetsRatesController: {
    getMessenger: getMultichainAssetsRatesControllerMessenger,
  },
  MultichainBalancesController: {
    getMessenger: getMultichainBalancesControllerMessenger,
  },
  ///: END:ONLY_INCLUDE_IF
} as const;
