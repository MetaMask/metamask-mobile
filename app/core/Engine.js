import {
	AccountTrackerController,
	AddressBookController,
	AssetsContractController,
	AssetsController,
	AssetsDetectionController,
	ComposableController,
	CurrencyRateController,
	KeyringController,
	PersonalMessageManager,
	MessageManager,
	NetworkController,
	NetworkStatusController,
	PhishingController,
	PreferencesController,
	TokenBalancesController,
	TokenRatesController,
	TransactionController,
	TypedMessageManager
} from '@metamask/controllers';

import AsyncStorage from '@react-native-community/async-storage';

import Encryptor from './Encryptor';
import { toChecksumAddress } from 'ethereumjs-util';
import Networks from '../util/networks';
import AppConstants from './AppConstants';
import { store } from '../store';
import { renderFromTokenMinimalUnit, balanceToFiatNumber, weiToFiatNumber } from '../util/number';
import NotificationManager from './NotificationManager';
import contractMap from 'eth-contract-metadata';
import Logger from '../util/Logger';
import { LAST_INCOMING_TX_BLOCK_INFO } from '../constants/storage';

const OPENSEA_API_KEY = process.env.MM_OPENSEA_KEY;
const encryptor = new Encryptor();
let refreshing = false;
/**
 * Core controller responsible for composing other metamask controllers together
 * and exposing convenience methods for common wallet operations.
 */
class Engine {
	/**
	 * ComposableController reference containing all child controllers
	 */
	datamodel;

	/**
	 * Object containing the info for the latest incoming tx block
	 * for each address and network
	 */
	lastIncomingTxBlockInfo;

	/**
	 * Creates a CoreController instance
	 */
	constructor(initialState = {}) {
		if (!Engine.instance) {
			const { nativeCurrency, currentCurrency } = initialState.CurrencyRateController || {
				nativeCurrency: 'eth',
				currentCurrency: 'usd'
			};
			this.datamodel = new ComposableController(
				[
					new KeyringController({ encryptor }, initialState.KeyringController),
					new AccountTrackerController(),
					new AddressBookController(),
					new AssetsContractController(),
					new AssetsController(),
					new AssetsDetectionController(),
					new CurrencyRateController({
						nativeCurrency,
						currentCurrency
					}),
					new PersonalMessageManager(),
					new MessageManager(),
					new NetworkController(
						{
							providerConfig: {
								static: {
									eth_sendTransaction: async (payload, next, end) => {
										const { TransactionController } = this.datamodel.context;
										try {
											const hash = await (await TransactionController.addTransaction(
												payload.params[0],
												payload.origin
											)).result;
											end(undefined, hash);
										} catch (error) {
											end(error);
										}
									}
								},
								getAccounts: (end, payload) => {
									const { approvedHosts, privacyMode } = store.getState();
									const isEnabled = !privacyMode || approvedHosts[payload.hostname];
									const { KeyringController } = this.datamodel.context;
									const isUnlocked = KeyringController.isUnlocked();
									const selectedAddress = this.datamodel.context.PreferencesController.state
										.selectedAddress;
									end(null, isUnlocked && isEnabled && selectedAddress ? [selectedAddress] : []);
								}
							}
						},
						{ network: '1', provider: { type: 'mainnet' } }
					),
					new NetworkStatusController(),
					new PhishingController(),
					new PreferencesController(
						{},
						{
							ipfsGateway: AppConstants.IPFS_DEFAULT_GATEWAY_URL
						}
					),
					new TokenBalancesController(),
					new TokenRatesController(),
					new TransactionController(),
					new TypedMessageManager()
				],
				initialState
			);

			const {
				AssetsController: assets,
				KeyringController: keyring,
				NetworkController: network,
				TransactionController: transaction
			} = this.datamodel.context;

			assets.setApiKey(OPENSEA_API_KEY);
			network.refreshNetwork();
			transaction.configure({ sign: keyring.signTransaction.bind(keyring) });
			network.subscribe(this.refreshNetwork);
			this.configureControllersOnNetworkChange();
			Engine.instance = this;
		}
		return Engine.instance;
	}

	configureControllersOnNetworkChange() {
		const {
			AccountTrackerController,
			AssetsContractController,
			AssetsDetectionController,
			NetworkController: { provider },
			TransactionController
		} = this.datamodel.context;

		provider.sendAsync = provider.sendAsync.bind(provider);
		AccountTrackerController.configure({ provider });
		AccountTrackerController.refresh();
		AssetsContractController.configure({ provider });
		TransactionController.configure({ provider });
		TransactionController.hub.emit('networkChange');
		AssetsDetectionController.detectAssets();
	}

	/**
	 * Refreshes all controllers that depend on the network
	 */
	refreshNetwork = () => {
		if (!refreshing) {
			refreshing = true;
			setTimeout(() => {
				this.configureControllersOnNetworkChange();
				refreshing = false;
			}, 500);
		}
	};

	refreshTransactionHistory = async forceCheck => {
		const { TransactionController, PreferencesController, NetworkController } = this.datamodel.context;
		const { selectedAddress } = PreferencesController.state;
		const { type: networkType } = NetworkController.state.provider;
		const { networkId } = Networks[networkType];
		try {
			const lastIncomingTxBlockInfoStr = await AsyncStorage.getItem(LAST_INCOMING_TX_BLOCK_INFO);
			const allLastIncomingTxBlocks =
				(lastIncomingTxBlockInfoStr && JSON.parse(lastIncomingTxBlockInfoStr)) || {};
			let blockNumber = null;
			if (
				allLastIncomingTxBlocks[`${selectedAddress}`] &&
				allLastIncomingTxBlocks[`${selectedAddress}`][`${networkId}`]
			) {
				blockNumber = allLastIncomingTxBlocks[`${selectedAddress}`][`${networkId}`].blockNumber;
				// Let's make sure we're not doing this too often...
				const timeSinceLastCheck = allLastIncomingTxBlocks[`${selectedAddress}`][`${networkId}`].lastCheck;
				const delta = Date.now() - timeSinceLastCheck;
				if (delta < AppConstants.TX_CHECK_MAX_FREQUENCY && !forceCheck) {
					return false;
				}
			} else {
				allLastIncomingTxBlocks[`${selectedAddress}`] = {};
			}
			//Fetch txs and get the new lastIncomingTxBlock number
			const newlastIncomingTxBlock = await TransactionController.fetchAll(selectedAddress, {
				blockNumber,
				alethioApiKey: process.env.MM_ALETHIO_KEY
			});
			// Check if it's a newer block and store it so next time we ask for the newer txs only
			if (
				allLastIncomingTxBlocks[`${selectedAddress}`][`${networkId}`] &&
				allLastIncomingTxBlocks[`${selectedAddress}`][`${networkId}`].blockNumber !== newlastIncomingTxBlock &&
				newlastIncomingTxBlock &&
				newlastIncomingTxBlock !== blockNumber
			) {
				allLastIncomingTxBlocks[`${selectedAddress}`][`${networkId}`] = {
					blockNumber: newlastIncomingTxBlock,
					lastCheck: Date.now()
				};

				NotificationManager.gotIncomingTransaction(newlastIncomingTxBlock);
			} else {
				allLastIncomingTxBlocks[`${selectedAddress}`][`${networkId}`] = {
					...allLastIncomingTxBlocks[`${selectedAddress}`][`${networkId}`],
					lastCheck: Date.now()
				};
			}
			await AsyncStorage.setItem(LAST_INCOMING_TX_BLOCK_INFO, JSON.stringify(allLastIncomingTxBlocks));
		} catch (e) {
			Logger.log('Error while fetching all txs', e);
		}
	};

	getTotalFiatAccountBalance = () => {
		const {
			CurrencyRateController,
			PreferencesController,
			AccountTrackerController,
			AssetsController,
			TokenBalancesController,
			TokenRatesController
		} = this.datamodel.context;
		const { selectedAddress } = PreferencesController.state;
		const { conversionRate } = CurrencyRateController.state;
		const { accounts } = AccountTrackerController.state;
		const { tokens } = AssetsController.state;
		let ethFiat = 0;
		let tokenFiat = 0;
		if (accounts[selectedAddress]) {
			ethFiat = weiToFiatNumber(accounts[selectedAddress].balance, conversionRate);
		}
		if (tokens.length > 0) {
			const { contractBalances: tokenBalances } = TokenBalancesController.state;
			const { contractExchangeRates: tokenExchangeRates } = TokenRatesController.state;
			tokens.forEach(item => {
				const exchangeRate = item.address in tokenExchangeRates ? tokenExchangeRates[item.address] : undefined;
				const tokenBalance =
					item.balance ||
					(item.address in tokenBalances
						? renderFromTokenMinimalUnit(tokenBalances[item.address], item.decimals)
						: undefined);
				const tokenBalanceFiat = balanceToFiatNumber(tokenBalance, conversionRate, exchangeRate);
				tokenFiat += tokenBalanceFiat;
			});
		}

		const total = ethFiat + tokenFiat;
		return total;
	};

	/**
	 * Returns true or false whether the user has funds or not
	 */
	hasFunds = () => {
		try {
			const {
				engine: { backgroundState }
			} = store.getState();
			const collectibles = backgroundState.AssetsController.collectibles;
			const tokens = backgroundState.AssetsController.tokens;
			const tokenBalances = backgroundState.TokenBalancesController.contractBalances;

			let tokenFound = false;
			tokens.forEach(token => {
				if (tokenBalances[token.address] && !tokenBalances[token.address].isZero()) {
					tokenFound = true;
				}
			});

			const fiatBalance = this.getTotalFiatAccountBalance();

			return fiatBalance > 0 || tokenFound || collectibles.length > 0;
		} catch (e) {
			Logger.log('Error while getting user funds', e);
		}
	};

	resetState = async () => {
		// Whenever we are gonna start a new wallet
		// either imported or created, we need to
		// get rid of the old data from state
		const {
			TransactionController,
			AssetsController,
			TokenBalancesController,
			TokenRatesController
		} = this.datamodel.context;

		//Clear assets info
		AssetsController.update({
			allCollectibleContracts: {},
			allCollectibles: {},
			allTokens: {},
			collectibleContracts: [],
			collectibles: [],
			ignoredCollectibles: [],
			ignoredTokens: [],
			suggestedAssets: [],
			tokens: []
		});

		TokenBalancesController.update({ contractBalances: {} });
		TokenRatesController.update({ contractExchangeRates: {} });

		TransactionController.update({
			internalTransactions: [],
			methodData: {},
			transactions: []
		});
	};

	sync = async ({ accounts, preferences, network, transactions, seed, pass, importedAccounts }) => {
		const {
			KeyringController,
			PreferencesController,
			NetworkController,
			TransactionController,
			AssetsController
		} = this.datamodel.context;

		// Select same network ?
		await NetworkController.setProviderType(network.provider.type);

		// Recreate accounts
		await KeyringController.createNewVaultAndRestore(pass, seed);
		for (let i = 0; i < accounts.hd.length - 1; i++) {
			await KeyringController.addNewAccount();
		}

		// Recreate imported accounts
		if (importedAccounts) {
			for (let i = 0; i < importedAccounts.length; i++) {
				await KeyringController.importAccountWithStrategy('privateKey', [importedAccounts[i]]);
			}
		}
		// Sync tokens
		const allTokens = {};
		Object.keys(preferences.accountTokens).forEach(address => {
			const checksummedAddress = toChecksumAddress(address);
			allTokens[checksummedAddress] = {};
			Object.keys(preferences.accountTokens[address]).forEach(
				networkType =>
					(allTokens[checksummedAddress][networkType] =
						networkType !== 'mainnet'
							? preferences.accountTokens[address][networkType]
							: preferences.accountTokens[address][networkType]
									.filter(({ address }) =>
										contractMap[toChecksumAddress(address)]
											? contractMap[toChecksumAddress(address)].erc20
											: true
									)
									.map(token => ({ ...token, address: toChecksumAddress(token.address) })))
			);
		});
		await AssetsController.update({ allTokens });

		// Restore preferences
		const updatedPref = { ...preferences, identities: {} };
		Object.keys(preferences.identities).forEach(address => {
			const checksummedAddress = toChecksumAddress(address);
			if (accounts.hd.includes(checksummedAddress) || accounts.simpleKeyPair.includes(checksummedAddress)) {
				updatedPref.identities[checksummedAddress] = preferences.identities[address];
			}
		});
		await PreferencesController.update(updatedPref);

		if (accounts.hd.includes(toChecksumAddress(updatedPref.selectedAddress))) {
			PreferencesController.setSelectedAddress(updatedPref.selectedAddress);
		} else {
			PreferencesController.setSelectedAddress(accounts.hd[0]);
		}

		await TransactionController.update({
			transactions: transactions.map(tx => ({
				id: tx.id,
				networkID: tx.metamaskNetworkId,
				origin: tx.origin,
				status: tx.status,
				time: tx.time,
				transactionHash: tx.hash,
				rawTx: tx.rawTx,
				transaction: {
					from: tx.txParams.from,
					to: tx.txParams.to,
					nonce: tx.txParams.nonce,
					gas: tx.txParams.gas,
					gasPrice: tx.txParams.gasPrice,
					value: tx.txParams.value
				}
			}))
		});

		return true;
	};
}

let instance;

export default {
	get context() {
		return instance && instance.datamodel && instance.datamodel.context;
	},
	get state() {
		const {
			AccountTrackerController,
			AddressBookController,
			AssetsContractController,
			AssetsController,
			AssetsDetectionController,
			CurrencyRateController,
			KeyringController,
			PersonalMessageManager,
			NetworkController,
			NetworkStatusController,
			PreferencesController,
			PhishingController,
			TokenBalancesController,
			TokenRatesController,
			TransactionController,
			TypedMessageManager
		} = instance.datamodel.state;

		return {
			AccountTrackerController,
			AddressBookController,
			AssetsContractController,
			AssetsController,
			AssetsDetectionController,
			CurrencyRateController,
			KeyringController,
			PersonalMessageManager,
			NetworkController,
			NetworkStatusController,
			PhishingController,
			PreferencesController,
			TokenBalancesController,
			TokenRatesController,
			TransactionController,
			TypedMessageManager
		};
	},
	get datamodel() {
		return instance.datamodel;
	},
	getTotalFiatAccountBalance() {
		return instance.getTotalFiatAccountBalance();
	},
	hasFunds() {
		return instance.hasFunds();
	},
	resetState() {
		return instance.resetState();
	},
	sync(data) {
		return instance.sync(data);
	},
	refreshTransactionHistory(forceCheck = false) {
		return instance.refreshTransactionHistory(forceCheck);
	},
	init(state) {
		instance = new Engine(state);
		Object.freeze(instance);
		return instance;
	}
};
