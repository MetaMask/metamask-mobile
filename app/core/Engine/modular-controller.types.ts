import { BaseControllerMessenger, Controllers } from './types';
import { STATELESS_NON_CONTROLLER_NAMES } from './constants';

export type Controller = Controllers[keyof Controllers];

/** The supported controller names. */
export type ControllerName = keyof Controllers;

/** All controller types by name. */
export type ControllerByName = {
  [name in ControllerName]: Controllers[name];
};

/**
 * Persisted state for all controllers.
 * e.g. `{ TransactionController: { transactions: [] } }`.
 */
export type ControllerPersistedState = Partial<{
  [Name in Exclude<
    ControllerName,
    (typeof STATELESS_NON_CONTROLLER_NAMES)[number]
  >]: Partial<ControllerByName[Name]['state']>;
}>;

/**
 * Request to initialize and return a controller instance.
 * Includes standard data and methods not coupled to any specific controller.
 */
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type ControllerInitRequest = {
  /**
   * Base controller messenger for the client.
   * Used to generate controller and init messengers for each controller.
   */
  baseControllerMessenger: BaseControllerMessenger;

  /**
   * Retrieve a controller instance by name.
   * Throws an error if the controller is not yet initialized.
   *
   * @param name - The name of the controller to retrieve.
   */
  getController<Name extends ControllerName>(
    name: Name,
  ): ControllerByName[Name];

  /**
   * The full persisted state for all controllers.
   * Includes controller name properties.
   * e.g. `{ TransactionController: { transactions: [] } }`.
   */
  persistedState: ControllerPersistedState;
};

/**
 * Result of initializing a controller instance.
 */
export interface ControllerInitResult<ControllerType extends Controller> {
  /**
   * The initialized controller instance.
   */
  controller: ControllerType;
}

/**
 * Function to initialize a controller instance and return associated data.
 */
export type ControllerInitFunction<ControllerType extends Controller> = (
  request: ControllerInitRequest,
) => ControllerInitResult<ControllerType>;
