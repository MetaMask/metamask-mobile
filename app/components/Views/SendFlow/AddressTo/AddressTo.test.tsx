import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import { render } from '@testing-library/react-native';

import Engine from '../../../../core/Engine';
import SendFlowAddressTo from './';

Engine.init({});
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigation: {},
  }),
  createNavigatorFactory: () => ({}),
}));

const initialState = {
  settings: {},
  engine: {
    backgroundState: {
      AccountTrackerController: {
        accounts: {
          '0x0': {
            balance: 200,
          },
        },
      },
      AddressBookController: {
        addressBook: {},
      },
      PreferencesController: {
        selectedAddress: '0x0',
        identities: {
          '0x0': {
            address: '0x0',
            name: 'Account 1',
          },
        },
      },
    },
  },
};

const mockStore = configureMockStore();
const store = mockStore(initialState);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest
    .fn()
    .mockImplementation((callback) => callback(initialState)),
}));

describe('SendFlowAddressTo', () => {
  it('should render correctly', () => {
    const wrapper = render(
      <Provider store={store}>
        <SendFlowAddressTo
          inputRef={undefined}
          highlighted={false}
          addressToReady={false}
          toSelectedAddress={undefined}
          toSelectedAddressName={undefined}
          onSubmit={() => undefined}
          inputWidth={undefined}
          confusableCollectionArray={undefined}
          isFromAddressBook={undefined}
          updateParentState={undefined}
          onToSelectedAddressChange={undefined}
        />
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
