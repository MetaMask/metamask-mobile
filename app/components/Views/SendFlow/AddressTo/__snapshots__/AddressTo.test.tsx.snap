// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendFlowAddressTo should render correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "marginHorizontal": 8,
      "minHeight": 52,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "width": "15%",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#24272A",
            "fontFamily": "EuclidCircularB-Regular",
            "fontSize": 30,
            "fontWeight": "400",
            "marginVertical": 2,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          Object {
            "color": "#24272A",
            "fontFamily": "EuclidCircularB-Regular",
            "fontSize": 16,
            "fontWeight": "400",
          },
        ]
      }
    >
      To:
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 8,
          "borderWidth": 1,
          "flex": 1,
          "flexDirection": "row",
          "marginLeft": 8,
          "marginVertical": 8,
          "minHeight": 52,
          "paddingHorizontal": 10,
        },
        Object {
          "borderColor": "#BBC0C5",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        autoFocus={true}
        keyboardAppearance="light"
        numberOfLines={1}
        onBlur={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Search, public address (0x), or ENS"
        placeholderTextColor="#BBC0C5"
        spellCheck={false}
        style={
          Array [
            Object {
              "color": "#24272A",
              "flex": 1,
              "fontFamily": "EuclidCircularB-Regular",
              "fontWeight": "400",
              "paddingLeft": 0,
              "paddingRight": 6,
            },
            undefined,
          ]
        }
        testID="txn-to-address-input"
      />
    </View>
    <TouchableOpacity
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 20,
            },
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "column",
              },
              Object {
                "color": "#24272A",
              },
            ],
            Object {
              "fontFamily": "anticon",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        îš—
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;
